### PR TITLE
37 add support to query local copies of a codebase rather than relying on githubs apis

### PR DESCRIPTION
**Pull request by tnzmnjm**
Hi Julian,

Could you please review the code and let me know how you'd like me to proceed?

Thanks

### CODE REVIEW

File: src/github_repo_request_local.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import subprocess
```
Comment (julianharty) [inline src/github_repo_request_local.py:1]: Generally I like us to include some file level documentation that briefly explains why this code exists.

Please can you add a docstring """ .... """ to describe this file's purpose?

File: src/github_repo_request_local.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import pandas as pd
+from loguru import logger
+import shutil
+
+
+def count_and_list_test_files(directory):
+    logger.info(f"Processing the directory {directory}")
+    test_files = []
+    # Define a set of file extensions to exclude
+    excluded_extensions = {
```
Comment (julianharty) [inline src/github_repo_request_local.py:12]: Good, as I've mentioned let's make this configurable at some point soon so that we don't need to edit code to change the exclusion criteria. For example if someone wants to exclude svg files, currently they'd need to format the exclusion correctly here to do so rather than editing a flat text file.

File: src/github_repo_request_local.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import subprocess
+from pathlib import Path
+import pandas as pd
+from loguru import logger
+import shutil
+
+
+def count_and_list_test_files(directory):
```
Comment (julianharty) [inline src/github_repo_request_local.py:8]: How about making this function simpler by getting it to do one thing - provide a list of the test files? the count can be performed trivially on the list that's returned by the calling code (e.g. line 66 here).

File: src/github_repo_request_local.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        ".txt",
+        ".md",
+        ".h",
+        ".xml",
+        ".html",
+        ".json",
+        ".png",
+        ".jpg",
+    }
+    for path in directory.rglob("*"):
```
Comment (julianharty) [inline src/github_repo_request_local.py:22]: Are these paths that are returned by `directory.rglob(...)`? It might be clearer to use the term `item` as both filenames and paths can be returned. See https://realpython.com/get-all-files-in-directory-python/#:~:text=To%20recursively%20list%20the%20items,their%20subdirectories%2C%20and%20so%20on.&text=.py')%5D-,The%20.,from%20the%20Path%20object%20recursively.

File: src/github_repo_request_local.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        if path.is_file() and (
+            "test" in path.name.lower() or "test" in str(path.parent).lower()
+        ):
+            # Exclude files with certain extensions
+            if path.suffix not in excluded_extensions:
+                test_files.append(str(path))
+    return len(test_files), test_files
+
+
+updated_csv_path = Path("../data/updated_local_github_df_test_count.csv")
```
Comment (julianharty) [inline src/github_repo_request_local.py:33]: If practical, please can you modify the code to use `git_codebase_root(...)` rather than a hard-coded path. This would make the code less dependent on the folder it's run from.

File: src/github_repo_request_local.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+if updated_csv_path.exists():
+    logger.info("Resuming from previously saved progress.")
+    df = pd.read_csv(updated_csv_path)
+else:
+    csv_file_path = Path("../data/local_github_df_test_count.csv")
```
Comment (julianharty) [inline src/github_repo_request_local.py:38]: ditto, please use `git_codebase_root(...)` if practical.

File: src/github_repo_request_local.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    df = df.drop("Unnamed: 0", axis=1)
+    df["testfilecountlocal"] = -1  # Initialize if first run
+
+for index, row in df.iterrows():
+    if row["testfilecountlocal"] != -1:
+        continue  # Skip processed repositories
+
+    repo_url = row["repourl"]
+    repo_name = Path(repo_url.split("/")[-1]).stem
+    clone_dir = Path.home() / "data" / "cloned_repos" / repo_name
```
Comment (julianharty) [inline src/github_repo_request_local.py:49]: This hardcoded path is good enough to get us started; in future it might be worth making this configurable e.g. so the cloned repos could be created on a separate volume with lots of available space rather than in the home folder's drive.

File: src/github_repo_request_local.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            continue
+
+    count, test_file_names = count_and_list_test_files(clone_dir)
+    df.at[index, "testfilecountlocal"] = count
+    formatted_names = "\n".join(test_file_names)
+    logger.info(f"Test file names for {repo_name}:\n{formatted_names}")
+
+    # Save after each update
+    df.to_csv(updated_csv_path, index=False)
+    shutil.rmtree(clone_dir)  # Clean up
```
Comment (julianharty) [inline src/github_repo_request_local.py:72]: As mentioned separately, we might want to make the removal optional.

**Review by julianharty — CHANGES_REQUESTED** [review]
It's looking good. I've added some suggestions that might be worth doing before we incorporate this functionality into the main codebase.

I suggest we leave the excluded_extensions configuration until another commit. The rest of my suggestions are probably worth doing now.

File: src/github_repo_request_local.py (hunk 1 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py:1]: `Docstring` has been added and committed.

File: src/github_repo_request_local.py (hunk 2 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py:18]: Could you please advice if using `argparse` module is a good approach or would it be better if we prompt the user to enter the file extensions to exclude at runtime?

File: src/github_repo_request_local.py (hunk 4 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - It returns all the files and folders from the Path object recursively.
- Change the loop variable in the function count_and_list_test_files from `path` to `item` for accuracy
- changes have been committed

File: src/github_repo_request_local.py (hunk 5 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - Added the function `git_codebase_root` to make the script less dependent on the folder it's   
run from and more versatile.

- It returns the absolute path of the top-level directory of the current Git repository. If
not in a Git repository, returns None.

- committed and pushed the code

File: src/github_repo_request_local.py (hunk 6 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - Refine `git_codebase_root` function to default to the current working directory if not within a Git repository, ensuring consistent path generation
- committed and pushed the code

File: src/github_repo_request_local.py (hunk 7 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py:77]: When required, Python's `argparse` module can be used or get the clone directory from an environment variable, with a default fallback.

File: src/github_repo_request_local.py (hunk 8 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - Implemented the function `should_keep_clones` for an interactive prompt for cleanup decision in the script
- Add a user prompt at the start of the script execution to determine whether cloned repositories should be deleted after being processed or kept
- changes have been committed and pushed to the branch

File: src/github_repo_request_local.py (hunk 7 above)
Comment (julianharty) [reply src/github_repo_request_local.py:94]: by environment variable do you mean one that's set using `export ENV_VAR=/home/julian/...` and/or one that's set using a command-line option such as `--working-dir=/home/julian/...`? I agree that either could be used especially if argparse supports both. If we only support one of these I'd prefer us to use command-line options for the working directory/folder as it's easier to see the setting when the program is run.

Comment (julianharty) [reply src/github_repo_request_local.py:94]: FYI I'm for using environment variables for secrets such as passwords rather than needing these to be passed as command line arguments where they're easier to be inadvertently logged.

File: src/github_repo_request_local.py (hunk 3 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: Soryy I missed this :)
- Changed the name of the function from `count_and_list_test_files(clone_dir)` to 
'list_test_files(clone_dir)'

File: src/github_repo_request_local.py (hunk 2 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - Implemented `argparse` for dynamic exclusion of file extensions, allowing users to specify which files to ignore when counting test files. 
- Logging the excluded file extensions `logger.info(f"Excluded file extensions: {', '.join(args.exclude)}")`
- code has been committed and pushed to the branch

File: src/github_repo_request_local.py (hunk 7 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: From my understanding both can be implemented . Can we leave environment variables for secrets as you mentioned? Because I also have used them in the same manner. I will resolve this by using the command-line option 

- Enabled configuration of the clone directory path through the --clone-dir command-line argument. This adjustment makes the script more flexible, accommodating environments where the default clone location is not suitable or preferred.
- Code has been pushed to the branch

File: src/github_repo_request_local.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        if item.is_file() and (
+            "test" in item.name.lower() or "test" in str(item.parent).lower()
+        ):
+            # Exclude files with certain extensions
+            if item.suffix not in excluded_extensions:
+                test_files.append(str(item))
+    return test_files
+
+
+def git_codebase_root():
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - Revert git_codebase_root to return None when not in a Git repository, aligning the function's behaviour with its purpose.
- Introduce a wrapper function `get_working_directory_or_git_root` to fallback to the current working directory when outside a Git repo, ensuring the script's flexibility and robustness.
- created the `utils` directory and inside it `git_utils.py` and put the ` git_codebase_root` and `get_working_directory_or_git_root` function to reduce code duplication and promote reuse across the codebase.
- Code pushed to the branch

File: src/github_repo_request_local.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            ["git", "rev-parse", "--show-toplevel"], stderr=subprocess.DEVNULL
+        )
+        return Path(root.decode().strip())
+    except subprocess.CalledProcessError:
+        logger.warning("Not inside a Git repository. Defaulting to current directory.")
+        return Path.cwd()
+
+
+def should_keep_clones():
+    """Prompt the user to decide whether to keep cloned repositories."""
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - Added non-interactive keep-clones command-line option to replace interactive prompt previously
implemented by the function `should_keep_clones()`
- code pushed to the branch

File: src/github_repo_request_local.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        if response in ("y", "yes"):
+            return True
+        elif response in ("n", "no"):
+            return False
+        else:
+            print("Please answer with 'y' or 'n'.")
+
+
+# Use git_codebase_root to define paths relative to the repository root
+repo_root = git_codebase_root()
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: I have added a `--keep-clones` flag that, when specified, indicates that the user wants to keep the cloned repositories. If the flag is not specified, the script will assume that the repositories should be removed after processing.
Please advice if more changes are required.

File: src/github_repo_request_local.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
+import subprocess
+from pathlib import Path
+import pandas as pd
+from loguru import logger
+import shutil
+
+"""
+This script automates the process of cloning GitHub repositories listed in a CSV file,
+counts the number of test files in each repository , and saves the count back to the CSV. It's designed to handle interruptions by saving progress
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: - These are resolved. code has been pushed to the branch.

File: src/github_repo_request_local.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        type=str,
+        default=str(Path.home() / "data" / "cloned_repos"),
+        help="Directory to clone repositories into. Defaults to "
+        "~/data/cloned_repos.",
+    )
+    parser.add_argument(
+        "--keep-clones",
+        action="store_true",
+        help="Keep cloned repositories after processing. If not specified, "
+        "cloned repositories ill be deleted.",
```
Comment (julianharty) [inline src/github_repo_request_local.py:42]: ...will be deleted :)

File: src/github_repo_request_local.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+            if item.suffix not in excluded_extensions:
+                test_files.append(str(item))
+    return test_files
+
+
+args = parse_args()
+# Log the excluded file extensions
+logger.info(f"Excluded file extensions: {', '.join(args.exclude)}")
+
+# Use git_codebase_root to define paths relative to the repository root
```
Comment (julianharty) [inline src/github_repo_request_local.py:68]: This comment is slightly out of date now you're calling the new function, just change the name of the function in the comment to fix.

File: utils/git_utils.py
Comment (julianharty) [inline utils/git_utils.py:1]: We'll want to update the rest of the code that calls `git_codebase_root(...)` so that they call this version of the function. Please can you do that as a separate PR?

File: utils/git_utils.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
+from pathlib import Path
```
Comment (julianharty) [inline utils/git_utils.py:1]: A minor suggestion re documenting this file - for the moment there's no need to document the file as both functions are documented and the file is tiny. At some point if/when the file contains more code it'll be good to document the file to summarise why it's needed, as well as the individual functions.

**Comment by julianharty** [comment]
@tnzmnjm Please merge once you've tidied up the remaining very minor changes.

File: src/github_repo_request_local.py (hunk 13 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py:42]: sorry. This is fixed now.

File: src/github_repo_request_local.py (hunk 14 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py:68]: This is fixed.

File: utils/git_utils.py (hunk 15 above)
Comment (tnzmnjm) [reply utils/git_utils.py:1]: - Added docstrings for the `utils/git_utils.py` file and its functions 
- pushed the code
